Source/Dafny/Resolver.cs

```diff
@@ -15397,16 +15400,23 @@ public class ResolveOpts {
       } else if (expr is StmtExpr) {
         var e = (StmtExpr)expr;
         int prevErrorCount = reporter.Count(ErrorLevel.Error);
+
+        var previousInOneStateStatementContext = inOneStateStatementContext;
+        if (!IsTwoStateContext(opts)) {
```

Comment:
Have you considered doing something like what `ResolveOpts` is for expressions, but then for statements?

```c#
record ResolveStatementContext(bool IsOneState, ICodeContext CodeContext);
public void ResolveStatement(Statement stmt, ResolveStatementContext context); 
```

instead of having the field `inOneStateStatementContext` in `Resolver`

That way you can't forget to restore the previous `inOneStateStatementContext` state, and it might make it easier to parallelise things in the future.

Reply:
Yes, that would be a good idea. Currently, options for statement translation uses these "global" fields and options for expression translation are passed as a parameter. However, I would suggest we do such a refactoring as a separate PR.

Reply:
If you leave the current global options for Statement as they are, and only make the change I suggested, would you still consider it a refactoring? To me it seems like a choice between adding a parameter to a (recursive)method or introducing a globalish variable, and then I prefer the former.

Reply:
Thanks for the suggestion! I did the change as new commits of this PR. This cleaned up a lot of stuff in how this was done before. Now, there's a `ResolutionContext` that gets passed through both statements and expressions.

Reply:
Btw, `ResolutionContext` is what previously was called `ResolveOpts`, which previously had been used only for expression resolution. I expanded it to work with statements, too.